### ISSUE TITLE
Extremely high GPU memory usage for a simple architecture

### ISSUE THREAD

**Issue by kamilkk852** [pytorch/pytorch#13722]
## 🐛 Bug

When using a particular architecture, pytorch is throwing CUDA OOME much faster (with "batch_size" of 10k) than tensorflow (which runs smoothly with "batch_size" of 200k-500k).

## To Reproduce

I wrote two almost identical implementations of a problematic NN architecture in both pytorch and tensorflow.

Here is PyTorch version:
```
import torch
import torch.nn as nn
import torch.nn.functional as F

class Model(nn.Module):
    def __init__(self, **params):
        super(Model, self).__init__()
        self.windows = [1, 2, 3]

        self.region_emb_layers = nn.ModuleList([nn.Conv2d(1, 150,
                                                kernel_size=(window, 300), stride=1) \
                                                for window in self.windows])
        self.beg_conv_layers = nn.ModuleList([nn.Conv2d(450, 450, \
                                              kernel_size=(3, 1), stride=1) \
                                              for _ in range(2)])                                   

        self.pooling = nn.MaxPool2d(kernel_size=(3, 1), stride=2)
        self.padding_conv = nn.ZeroPad2d((0, 0, 0, 2))
        self.act_fun = nn.ReLU()
        self.linear_out = nn.Linear(450, 2)

    def forward(self, out, emb):
        outs = [None]*len(self.windows)
        out = emb(out)
        out = out.unsqueeze(1)
        
        for i, (window, region_emb_layer) in enumerate(zip(self.windows, self.region_emb_layers)):
            out_ = region_emb_layer(out)
            out_ = nn.ZeroPad2d((0, 0, 0, window - 1))(out_)
            outs[i] = out_

        out = torch.cat(outs, dim=1)
        
        for beg_conv_layer in self.beg_conv_layers:
            out = self.padding_conv(out)
            out = beg_conv_layer(out)
            out = self.act_fun(out)
            
        out = F.adaptive_max_pool2d(out, (1, 1))
        out = out.view(out.size(0), -1)
        out = self.linear_out(out)
        
        return out

import numpy as np

def gen(dataset_size, batch_size):
    for i in range(dataset_size // batch_size):
        b = np.random.choice([10, 30, 70, 200, 500, 1000])
        a = batch_size // b

        x = np.random.randint(0, 42000, (a, b))
        y = np.random.randint(0, 2, (a,))
        
        yield x, y


device = torch.device('cuda:0')

model = Model().to(device)
model = model.to(device)

print('Model parameters cnt: {}'.format(sum(p.numel() for p in model.parameters())))

emb = nn.Embedding(42000, 300).to(device)
emb.weight.requires_grad_(False)

optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)

dataset_size = int(1e6)
batch_size = 10000

for i, (x, y) in enumerate(gen(dataset_size, batch_size)):
    if i % 10 == 0:
        print(i)

    x = torch.from_numpy(x).to(device)
    y = torch.from_numpy(y).to(device)

    out = model(x, emb=emb)

    crit = nn.CrossEntropyLoss()
    loss = crit(out, y)
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
```

And tensorflow:
```
import tensorflow
import numpy as np
import keras
from keras.models import Model
from keras.layers import Lambda, Input, Conv2D, Dense, Flatten, \
Embedding, Concatenate, ZeroPadding2D
from keras.optimizers import Adam
import keras.backend as K


x_in = Input((None,))

x = x_in

x = Embedding(42000, 300, trainable=False)(x)
x = Lambda(lambda x: K.expand_dims(x, axis=3))(x)

x1 = ZeroPadding2D(((0, 2), (0, 0)))(x)
x1 = Conv2D(150, (3, 300), activation='relu')(x1)
x2 = ZeroPadding2D(((0, 1), (0, 0)))(x)
x2 = Conv2D(150, (2, 300), activation='relu')(x2)
x3 = Conv2D(150, (1, 300), activation='relu')(x)

x = Concatenate(axis=3)([x1, x2, x3])

x = Conv2D(450, (3, 1), activation='relu', padding='same')(x)
x = Conv2D(450, (3, 1), activation='relu', padding='same')(x)

x = Lambda(lambda x: K.max(x, axis=1))(x)

x = Flatten()(x)
x = Dense(1, activation='sigmoid')(x)

x_out = x

model = Model(inputs=x_in, outputs=x_out)
opt = Adam(lr=1e-3)
model.compile(loss='binary_crossentropy', optimizer=opt)

print(model.summary())

def gen(batch_size):
  while True:
    b = np.random.choice([10, 30, 70, 200, 500, 1000])
    a = batch_size // b
    x = np.random.randint(0, 42000, (a, b))
    y = np.random.randint(0, 2, (a,))
    
    yield x, y

dataset_size = int(1e6)
batch_size = 200000

model.fit_generator(gen(batch_size), steps_per_epoch=dataset_size//batch_size)
```
If you run both scripts on 11GB GPU you should see CUDA OOME in pytorch (on loss.backward() line), but not in tensorflow script.

## Expected behavior

I would expect adequate GPU memory usage for such a simple architecture and small input size (10k numbers), which is the case in tensorflow, but not in pytorch.

## Environment

(This is the environment of Google Colaboratory, but the same issue on two other machines)

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.2.148

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.2.148
GPU models and configuration: GPU 0: Tesla K80
Nvidia driver version: 396.44
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.3.1

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

**Comment by zou3519** [comment]
We have fixed this significantly on master; please try building from source.

**Comment by kamilkk852** [comment]
Thanks, it worked!
